### PR TITLE
[codex] Add ToggleGroup agnostic core

### LINKED ISSUE #216: task: Implement ToggleGroup agnostic core
**Epic: #10 (Agnostic utility components)**

## Goal

Implement the framework-agnostic ToggleGroup component state machine: the most complex utility component. Idle/Focused states with item tracking, SelectionMode (None/Single/Multiple), roving tabindex keyboard navigation (ArrowLeft/Right/Up/Down, Home/End), RTL-aware arrow key swapping, registered_items management, `Bindable<BTreeSet<Key>>` for controlled/uncontrolled multi-selection, per-item context distribution, and form integration via hidden inputs.

## Out of scope

Framework-specific adapter components (Leptos/Dioxus) and direct DOM focus/measurement calls.

## Spec refs

- `spec/components/utility/toggle-group.md`
- `spec/foundation/06-collections.md`

## Depends on

#209 (Toggle agnostic core), #215 (ToggleButton agnostic core)

## Layer

Component

## Framework

None

## Test tier

Unit

## Fibonacci points

8

## Element/ref handling note

- Core should register items by stable values/keys and maintain logical focus order.
- Roving focus outputs should be attrs plus adapter-resolvable focus intent keyed by item value, not a DOM ID lookup.
- Indicator positioning or other geometry should be represented as logical state/intent only; adapters own live root/item/indicator handles and measurements.
- IDs remain valid for form, ARIA, and testable attributes, but not as the primary way to focus or measure items.

## Tests to add first

- State machine: initial state is Idle with no selection (default_value).
- State machine: Focus event transitions Idle→Focused.
- State machine: SelectItem/DeselectItem events update selection.
- State machine: Single mode allows only one selected item.
- State machine: Multiple mode allows multiple selected items.
- Keyboard navigation: ArrowRight moves focus to next item.
- Keyboard navigation: ArrowLeft moves focus to previous item.
- Keyboard navigation: Home/End move to first/last item.
- Keyboard navigation: RTL mode swaps ArrowLeft/ArrowRight behavior.
- Keyboard navigation: roving tabindex — only focused item has tabindex=0.
- Item registration: RegisterItem/UnregisterItem maintain item list.
- ConnectApi Root: produces `role="group"`, `aria-label`, `data-ars-orientation`.
- ConnectApi Item: produces `aria-pressed`, `tabindex`, `data-ars-state` per item.
- Controlled mode: `value` prop overrides internal selection.
- Bindable: on_value_change fires on selection changes.
- Form integration: hidden inputs for each selected value when name is set.
- Props: `value`, `default_value`, `selection_mode`, `orientation`, `disabled`, `loop`, `roving_focus` match spec.

## Acceptance criteria

- All public types documented per workspace `missing_docs` lint.
- State machine matches spec §1 exactly.
- ConnectApi produces correct ARIA attributes per spec §2 for Root and Item parts.
- Roving tabindex keyboard navigation with RTL support works through stable item identity and adapter-resolvable focus intent.
- Selection modes (None/Single/Multiple) enforced.
- Form integration with hidden inputs matches spec.
- Snapshot tests cover the component's `connect()` / `Api` `AttrMap` output for every anatomy part and each output-affecting state, prop, or context branch, with committed `.snap` fixtures passing `cargo insta test --unreferenced=reject`.
- All named tests pass with zero warnings.

## Spec impact
No spec change required unless current spec examples require ID-only focus/measurement. Fix those in the same task if found.

## Board update rule

If blocked, keep `Status` in `Todo`, use native GitHub issue dependencies, and name the dependency or spec ambiguity.

### PR DESCRIPTION
**Pull request by fogodev**
## Summary

- Add the framework-agnostic ToggleGroup core state machine, connect API, ARIA output, hidden input configuration, snapshots, and public utility export.
- Add unit, snapshot, spec-conformance, and proptest coverage for selection modes, controlled/uncontrolled value sync, item registration, roving focus, keyboard navigation, and hidden input output.
- Sync the ToggleGroup component spec and adapter specs for registration events, `ars_collections::Key`, adapter-resolvable focus intent, and controlled/uncontrolled prop behavior.

Closes #216

## Validation

- `cargo xci-fast`
- `cargo test -p ars-components toggle_group_`
- `cargo test -p ars-components --test spec_conformance utility::toggle_group_anatomy_matches_spec`
- `PROPTEST_CASES=256 cargo test -p ars-components --test proptest_state_machines utility::proptest_toggle_group_event_sequences_preserve_invariants -- --ignored`
- `cargo insta test --unreferenced=reject -p ars-components --features i18n --lib`
- `cargo xtask spec validate`
- `cargo xtask spec toc spec/components/utility/toggle-group.md`
- `cargo llvm-cov test -p ars-components --features i18n --lib --no-fail-fast --text -- utility::toggle_group`
- `cargo mutants -p ars-components -f crates/ars-components/src/utility/toggle_group/mod.rs --jobs 8`
- `cargo +nightly fmt --all --check`
- `cargo test -p ars-components --all-targets --features i18n`

### CODE REVIEW

**Comment by fogodev** [comment]
@codex review

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 1]
```diff
@@ -0,0 +1,2556 @@
...
+
+        if self.props.invalid {
+            attrs
+                .set(HtmlAttr::Aria(AriaAttr::Invalid), "true")
+                .set_bool(HtmlAttr::Data("ars-invalid"), true);
+        }
+
+        if self.props.required {
+            attrs.set(HtmlAttr::Aria(AriaAttr::Required), "true");
+        }
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs]: Fixed by gating `aria-required` to `SelectionMode::Single`, where the root uses `role="radiogroup"`. Added `toggle_group_group_role_root_attrs_omit_unsupported_aria` and updated the spec/snapshot.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 2]
```diff
@@ -0,0 +1,2556 @@
...
+
+        let selected = self.is_selected(item_id);
+        let focused = self.ctx.focused_item.as_ref() == Some(item_id);
+        let disabled = self.is_item_disabled(item_id);
+
+        attrs
+            .set(scope_attr, scope_val)
+            .set(part_attr, part_val)
+            .set(HtmlAttr::Data("ars-value"), item_id.to_string())
+            .set(HtmlAttr::TabIndex, self.item_tabindex(item_id, disabled));
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs]: Fixed by adding `HtmlAttr::Type = "button"` to item attrs. Covered in the single and multiple item attribute tests and refreshed item snapshots.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 3]
```diff
@@ -0,0 +1,2556 @@
...
+        let [(scope_attr, scope_val), (part_attr, part_val)] = Part::Root.data_attrs();
+
+        let orientation = orientation_token(self.ctx.orientation);
+
+        attrs
+            .set(scope_attr, scope_val)
+            .set(part_attr, part_val)
+            .set(HtmlAttr::Role, root_role(self.ctx.selection_mode))
+            .set(HtmlAttr::Aria(AriaAttr::Orientation), orientation)
+            .set(HtmlAttr::Data("ars-orientation"), orientation);
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:843]: Fixed by emitting `aria-orientation` only for `SelectionMode::Single` / `role="radiogroup"`. `data-ars-orientation` remains available for styling in all modes. Added regression coverage and updated the spec/snapshot.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 4]
```diff
@@ -0,0 +1,2556 @@
...
+        state: &Self::State,
+        event: &Self::Event,
+        ctx: &Self::Context,
+        props: &Self::Props,
+    ) -> Option<TransitionPlan<Self>> {
+        if ctx.disabled
+            && matches!(
+                event,
+                Event::SelectItem(_) | Event::DeselectItem(_) | Event::ToggleItem(_) | Event::Reset
+            )
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:516]: Fixed by removing `Event::Reset` from the disabled value-event guard. Added `toggle_group_reset_restores_default_value_when_disabled` and synced the spec reset wording.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 5]
```diff
@@ -0,0 +1,2556 @@
...
+        .cloned()
+}
+
+fn idle_focus_seed(ctx: &Context, step: FocusStep) -> Option<Key> {
+    if let Some(selected) = ctx
+        .value
+        .get()
+        .iter()
+        .find(|item| can_focus_item(ctx, item))
+        .cloned()
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:1180]: Fixed by seeding idle focus from `registered_items` order before falling back to first/last enabled item. Added `toggle_group_idle_focus_uses_registered_selected_order`.

**Comment by fogodev** [comment]
@codex review

File: crates/ars-components/tests/proptest_state_machines/utility.rs [hunk 6]
```diff
@@ -459,6 +558,126 @@ proptest! {
...
+            let before_read_only = service.props().read_only;
+
+            let value_item_event = matches!(
+                event,
+                toggle_group::Event::SelectItem(_)
+                    | toggle_group::Event::DeselectItem(_)
+                    | toggle_group::Event::ToggleItem(_)
+            );
+
+            let disabled_value_event = value_item_event || matches!(event, toggle_group::Event::Reset);
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/tests/proptest_state_machines/utility.rs]: Fixed by excluding `Reset` from the disabled item-value invariant. The proptest now only asserts disabled groups block item selection events, while reset remains allowed.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 7]
```diff
@@ -0,0 +1,2604 @@
...
+const fn orientation_token(orientation: Orientation) -> &'static str {
+    match orientation {
+        Orientation::Horizontal => "horizontal",
+        Orientation::Vertical => "vertical",
+    }
+}
+
+fn is_item_disabled(ctx: &Context, item: &Key) -> bool {
+    ctx.disabled || ctx.disabled_items.contains(item)
+}
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:1147]: Fixed by separating activation-disabled from focus-disabled checks. Group-disabled items still expose disabled semantics and block activation, but registered items remain roving-focus targets unless individually disabled. Added unit coverage for disabled-group focus navigation and prop sync.

**Comment by fogodev** [comment]
@codex review

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 8]
```diff
@@ -0,0 +1,2641 @@
...
+            KeyboardKey::ArrowRight if horizontal => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowLeft if horizontal && rtl => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowLeft if horizontal => (self.send)(Event::FocusPrev),
+            KeyboardKey::ArrowDown if !horizontal => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowUp if !horizontal => (self.send)(Event::FocusPrev),
+            KeyboardKey::Home => (self.send)(Event::FocusFirst),
+            KeyboardKey::End => (self.send)(Event::FocusLast),
+            KeyboardKey::Space | KeyboardKey::Enter => {
+                (self.send)(Event::ToggleItem(item_id.clone()));
+            }
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs]: Fixed in d8a1c2d4. `on_item_keydown` no longer dispatches `ToggleItem` for Space/Enter; activation now flows only through the native button click path, avoiding duplicate keyboard activation. I also updated the keydown tests and spec wording to make this boundary explicit.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 9]
```diff
@@ -0,0 +1,2641 @@
...
+
+        match data.key {
+            KeyboardKey::ArrowRight if horizontal && rtl => (self.send)(Event::FocusPrev),
+            KeyboardKey::ArrowRight if horizontal => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowLeft if horizontal && rtl => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowLeft if horizontal => (self.send)(Event::FocusPrev),
+            KeyboardKey::ArrowDown if !horizontal => (self.send)(Event::FocusNext),
+            KeyboardKey::ArrowUp if !horizontal => (self.send)(Event::FocusPrev),
+            KeyboardKey::Home => (self.send)(Event::FocusFirst),
+            KeyboardKey::End => (self.send)(Event::FocusLast),
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:1044]: Fixed in d8a1c2d4. `on_item_keydown` now returns early when `ctx.roving_focus` is false, and the regression test covers arrow/Home/End no-ops plus Space/Enter activation no-ops in that mode. The spec now states that roving key translation is gated on `roving_focus=true`.

File: crates/ars-components/src/utility/toggle_group/mod.rs [hunk 10]
```diff
@@ -0,0 +1,2641 @@
...
+            }
+
+            (_, Event::RegisterItem(item)) => {
+                if ctx
+                    .registered_items
+                    .iter()
+                    .any(|registered| registered == item)
+                {
+                    return None;
+                }
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/toggle_group/mod.rs:646]: I’m leaving this unchanged intentionally. The issue plan for #216 explicitly requires `RegisterItem(key)` to append only if not already present, and this PR has a regression test (`toggle_group_register_deduplicates_existing_keys`) for that contract. In this agnostic core, stable `Key` values are the adapter identity; changing duplicate registration to fail fast would be a task/spec contract change rather than a bug fix.

**Comment by fogodev** [comment]
@codex review
